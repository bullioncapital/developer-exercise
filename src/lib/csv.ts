import * as colors from 'colors';
import * as parser from 'csvtojson';
import * as fs from 'fs';

export const processor = (async(file: string) => {
    let output: any[] = [] ;

    console.log(colors.blue(`Initialise reading data from ${file}`));

    await parser()
    .fromFile(file)
    .then((jsonObj) => {
        output = jsonObj;
    })

  return output;
});
