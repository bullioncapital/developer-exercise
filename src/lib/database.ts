import { getModel } from './model/stats';

export const getFilterdStats = (async (indicator: string) => {
    const StatsModel: any = await getModel();
    const response: Array<object> = await StatsModel.find()
    .where('Indicator Name').equals(indicator)
    .catch((error) => {
      console.log('error : ', error);
      throw new Error(error);
    });

    return response
});