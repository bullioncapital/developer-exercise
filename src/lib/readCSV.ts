import * as colors from 'colors';
import * as parser from 'csvtojson/v2';

import * as csv from 'csv-parser';
import * as fs from 'fs';

export const processor = (async (file: string) => {
    let output: any[] = [] ;

    console.log(colors.blue(`Initialise reading data from ${file}`));

    await parser({ eol: '\n' })
    .fromFile(file)
    .then((jsonObj) => {
        output = jsonObj;
    })

  return output;
});
