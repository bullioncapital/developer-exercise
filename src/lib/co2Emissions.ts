import { getFilterdStats } from './database';

export interface IHighestAverage {
    country: string;
    average: number;
}

// exoected return result {"country":"World","average":21348295.636363637}
export const getCo2EmissionsAverage = (async (): Promise<IHighestAverage> => {
    let highestAverage: IHighestAverage = { country: '', average: 0 };
    const filterdData: Array<object> = await getFilterdStats('CO2 emissions (kt)');
    
    filterdData.map((row) => {
        let count: number = 0;
        let totalAvalable = 0;
        for (let i=1960; i<=2017; i++) {
            const value = row[`${i.toString()}`]
            if (value !== '') {
                totalAvalable++;
                count = count + parseInt(value);
            }
        }

        const currentAverage = count/totalAvalable;
        if (highestAverage.average < currentAverage) {
            highestAverage.country = row['Country Name'];
            highestAverage.average = parseFloat((currentAverage as any).toFixed(2));
        }
    });

    return highestAverage;
});
