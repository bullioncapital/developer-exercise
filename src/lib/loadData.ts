import { each as BPEach } from 'bluebird';
import { processor } from './readCSV';
import { getModel } from './model/stats';
import { statsFields } from './model/statsFields';

(async () => {
    const dataSet = await processor('src/lib/data/data.csv');
    const StatsModel: any = await getModel();

    let rowObject = {};
    await BPEach(dataSet, async (data) => {
        statsFields.map((value) => {
            rowObject[value] = data[value];
        });
        const Stats: any = new StatsModel(rowObject);
        const response = await Stats.save()
        .catch((error) => {
            console.log('Error: ', error);
        });

        console.log(response);
    });
    
    process.exit();
})();
