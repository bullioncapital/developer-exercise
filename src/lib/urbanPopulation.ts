import { IHighestAverage } from './co2Emissions';
import { getFilterdStats } from './database';
 
// expected return {"country":"Botswana","average":12.9}
export const getAverageUrbanPopulation = (async (): Promise<IHighestAverage> => {
    const filterdData: Array<object> = await getFilterdStats('Urban population growth (annual %)');

    let highestAverage: IHighestAverage = { country: '', average: 0 };
    filterdData.map((row) => {
        let fullRange: boolean = true;
        let count: number = 0;
        for (let i=1980; i<=1990; i++) {
            const value = row[`${i.toString()}`]
            if (value === '' || value === undefined) {
                fullRange = false;
            }
            count = count + parseInt(value);
        }

        if (fullRange) {
            const currentAverage = count/10;
            if (highestAverage.average < currentAverage) {
                highestAverage.country = row['Country Name'];
                highestAverage.average = currentAverage;
            }
        }
    });

    return highestAverage;
});