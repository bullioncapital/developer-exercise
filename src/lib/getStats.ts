import * as Ajv from 'ajv';
import { getAverageUrbanPopulation } from './urbanPopulation';
import { getCo2EmissionsAverage } from './co2Emissions';
import * as schema from './../../resources/schema/statsRequest.json';

export const getStats = (async (ctx) => {
    let response: any;
    const { request: { body } } = ctx;
    const { reportid } = body;

    // validate the request agains the json schema to ensure intergrity of the request data
    const ajv = new Ajv ({ allErrors: true });
    const valid = ajv.validate(schema, body); 
    if (!valid) {
        console.log('validation Errors : ', ajv.errors);
        // todo: make the errors json API complient.
        throw Error('Schema Validation Error!');
    }

    if (reportid === 1) {
        response = await getAverageUrbanPopulation();
    } else if ((reportid === 2)) {
        response = await getCo2EmissionsAverage();
    } else {
        response = { notice: 'report type not implemented!' };
    }

    if (ctx) {
        ctx.body = response;
    }

    return ctx;
});
