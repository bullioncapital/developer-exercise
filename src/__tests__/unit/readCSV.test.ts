import { processor } from '../../lib/readCSV';

describe('getAverageUrbanPopulation', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('is a function', () => {
        expect(processor).toBeInstanceOf(Function);
    });

    it('returns correct output', async () => {
        const response = await processor('src/__tests__/unit/fixtures/data.csv');

        expect(response).toBeTruthy();
        expect(response.length).toEqual(2);
    });

    it('returns read error', async () => {
        let errors;
        await processor('').catch((err) => errors = err);
    
        expect(errors).toBeTruthy();
        expect(errors).toThrowErrorMatchingSnapshot();
    });
});