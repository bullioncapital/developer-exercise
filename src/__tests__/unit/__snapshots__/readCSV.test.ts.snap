// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAverageUrbanPopulation returns read error 1`] = `"received is not a function"`;
